### ISSUE TITLE
Feature Request: Run Tests in Parallel with Shared Context

### ISSUE THREAD

**Issue by zijianhuang** [xunit/xunit#2055]
**Background**
Currently by default tests in the same assembly but in different classes run in parallel.

To run tests in different classes not in parallel, you put them  into the same test collection.

To have multiple classes sharing the same contexts, you use a Collection Fixture, however, then the tests run one by one.

I have some integration tests that need to launch IIS or DotNet Host, or drop/create databases, via collection fixtures.

**Feature Request**

Allow to run tests in different classes in parallel sharing the same context like dropping/creating databases. 

Programming interface expected:

```csharp
public sealed class CollectionAttribute : Attribute
	{
		public CollectionAttribute(string name, boolean parallel=false);
	} 


[Collection("Database collection", true)]
public class DatabaseTestClass1
{
 //...
}

[Collection("Database collection", true)]
public class DatabaseTestClass2
{
    // ...
}
```

**Comment by redoz** [comment]
I'll chime in and say it would be very nice if v3 offered more control over how tests in collections were executed (e.g. in parallel) or offered up a different means of having shared contexts and controlling execution. I understand this isn't really a concern when unit-testing, but we're using xbehave as our integration testing framework (mostly because it lets us re-use existing infrastructure that supports xunit, runners, reporting, etc.) and I've ended up hacking on my own "AssemblyFixture" support so we can handle shared context across our test-suite. But even so I'm forced to run some tests in a non-parallel collection simply to make sure it doesn't interfere with the rest of the test-suite which is a bit inefficient as those tests run serialized now.

**Comment by fjmorel** [comment]
I would love something like this. We currently use a CollectionFixture to start a Mongo server for the majority of our tests, but are using randomized collection names for each test to avoid having to clean up between each test. Since they can't conflict with each other, being able to run all the tests within this collection in parallel would help a lot.

**Comment by markm77** [comment]
I made a comment on the other parallelisation issue: https://github.com/xunit/xunit/issues/1986#issuecomment-669429834.

Yes, it would be really great to have the ability to run tests in parallel when using a collection fixture. We use Xunit for both unit and integration tests, and for the integration tests often instance a webhost via a collection fixture. In addition, we also often use a single `[Theory]` test with multiple `[MemberData]` cases (i.e. dynamic test case generation). I'm really hoping the implementation of this allows `[MemberData]` cases for a single `[Theory]` test to run in parallel!

**Comment by pdevito3** [comment]
This will be a **huge** win for my projects either way, but in the proposed API, why default these to not be parallelized? I'd rather have it be opt out than opt in and would think that would be a better default?

**Comment by bradwilson** [comment]
@pdevito3 Because that's how it works in v2, and by changing the default, it could potentially cause a bunch of tests to fail without knowing why when a user upgrades from v2 to v3. Additionally, I consider this to be an advanced feature, so enabling it makes sense to me.

**Comment by pdevito3** [comment]
Makes sense. As long as I can turn it on I'll be a happy camper 🙂

**Comment by pdevito3** [comment]
👋 @bradwilson just saw that v3's out but this didn't make it in :-( any thoughts on it's potential future?

**Comment by bradwilson** [comment]
This is still something that I would like to get into v3, it just didn't make the bar for 1.0.

I don't have a particular design or time frame in mind at this moment.

**Comment by dazinator** [comment]
Just sharing our use case for this feature:- 
1. We have a set of Repository classes, and each one has a test class that performs integration tests against a sql container.
2. The classes use a fixture and IClassFixture to have the test container setup at the start of the test. However this means they each get an isolated sql container. They don't need an isolated sql container because we have each test use an independent tenant ID when querying the dataabse so they can't step on each others toes. 
  - They don't need seperate databases or schema, they only need to ensure they don't impact each others data - which we accomplish with tenant id isolation
 
Tried to switch to a collection fixture and this enables a single "sql container" but it then runs them all sequentially against that resource - as they are isolated we want to allow them to run in parallel still.

**Comment by jonasspecht** [comment]
Just want to chime in and say that this would be a very nice feature. 
We have a similar use case to what others have already mentioned with a database in a test container which we would like to share across multiple test classes by using a collection fixture, while still keeping the parallelism between the classes (tests are separated by a tenant id).

**Comment by matthew-mabrey** [comment]
I'm thinking of attempting an implementation for this and https://github.com/xunit/xunit/issues/1986. My current idea is:
- Add an option `EnableTestCaseParallelization` to test collections / collection attributes so that it's opt in.
- Rework code on the path from `XunitTestCollectionRunner -> TestRunnerBase` to be similar to `XunitTestAssemblyRunnerBase.RunTestCollections` (adding tasks to a task collection when parallelization is enabled) since a lot of the code on that path linearly awaits the tasks for all the test classes in a test collection, test methods in a test class, and test cases for a test method.
- To maintain `ParallelAlgorithm.Conservative` behavior, pass an optional `SemaphoreSlim? parallelizationSemaphore` down all the way through to `TestRunnerBase.Run` via `ContextBase`. Collections with test case parallelization enabled will enter the semaphore at the start of `TestRunnerBase.Run`. Collections with the test case parallelization disabled will enter the semaphore in `TestCollectionRunner.RunTestClasses`.

@bradwilson thoughts when you get the chance? Just want to make sure I'm not missing anything or going against any core design decisions.

**Comment by bradwilson** [comment]
@mattmabrey1 On the surface, this seems reasonable, although I have a slightly uncomfortable feeling about "passing the semaphore down". I can't really say why, or what I might prefer (encapsulation, maybe?).

It's probably also worth updating `[CollectionBehavior]` to allow a global default override to `true`, to prevent people from having to put everything into collections just to get access to the parallelization flag (since I suspect most people who want this and/or #1986 to want a way to turn it on by default).

FWIW, I expect this to be a lot more work than you probably expect, given how much of the execution pipeline is going to have to be updated just for the propagation, and depending on how it all lands, may end up needing to be part of 4.0 due to breaking changes.

**Comment by matthew-mabrey** [comment]
> although I have a slightly uncomfortable feeling about "passing the semaphore down". I can't really say why, or what I might prefer (encapsulation, maybe?).

Yeah for now I'm thinking of just adding the semaphore as a property in `ContextBase` since that already gets passed down the execution pipeline. If that's not preferred, another option I can think of is having the semaphore as an internal static variable somewhere at the test assembly level since IIRC we are only defining one of those per test process. That seemed less flexible with future changes though (like if for whatever reason a future user wants to pass down different semaphores for different test collections/classes)

> It's probably also worth updating [CollectionBehavior] to allow a global default override to true

Ah good point, I'll do that 👍 

> FWIW, I expect this to be a lot more work than you probably expect, given how much of the execution pipeline is going to have to be updated just for the propagation, and depending on how it all lands, may end up needing to be part of 4.0 due to breaking changes.

Yeah I initially assumed this would be an easier change before I started digging through the xUnit internals more 😅 I think you're right though and it most likely will require breaking changes that push it to whatever the next major release is.

Thanks for the quick response and feedback!

**Comment by bradwilson** [comment]
> Yeah for now I'm thinking of just adding the semaphore as a property in `ContextBase` since that already gets passed down the execution pipeline

Yep, that's the correct place to put it. The context exists so that we don't have a whole bunch of 40 parameter methods. 😂

**Comment by bradwilson** [comment]
Full test parallelism is now available in v3 `4.0.0-pre.151` https://xunit.net/docs/using-ci-builds

Updated parallel mode documentation is available: https://xunit.net/docs/running-tests-in-parallel#parallel-modes

**Comment by zijianhuang** [comment]
To clarify, so far I don't see tag or branch 4.0.0-pre.151 but 4.0.0-pre.154.

I have checked out 4.0.0-pre.154, built and imported. The key is to declare in the test assembly:
```csharp
[assembly: Parallelization(Mode = ParallelMode.All)]
```

The console runner is adapting the new parallel mode All, while the visualstudio runner 4.0.0-pre.5 is not. I would assume bradwilson team is working on the visualstudio runner.

**Comment by bradwilson** [comment]
> The console runner is adapting the new parallel mode All, while the visualstudio runner 4.0.0-pre.5 is not. I would assume bradwilson team is working on the visualstudio runner.

I don't know what this means. Can you provide an example?

**Comment by bradwilson** [comment]
I think maybe the bug is that the wrong value is being shown in the diagnostic headers when using the attribute for parallelization, but it is being honored.

Example:

```csharp
using Xunit.Sdk;
using Xunit.v3;

[assembly: Parallelization(Mode = ParallelMode.All)]

public class UnitTest1
{
    [Fact]
    public void Test1()
    {
        TestContext.Current.SendDiagnosticMessage("Starting Test1");
        Thread.Sleep(3000);
        TestContext.Current.SendDiagnosticMessage("Finished Test1");
    }

    [Fact]
    public void Test2()
    {
        TestContext.Current.SendDiagnosticMessage("Starting Test2");
        Thread.Sleep(2000);
        TestContext.Current.SendDiagnosticMessage("Finished Test2");
    }
}
```

Output:

<img width="1016" src="https://github.com/user-attachments/assets/f7d1761d-0e1e-41c1-99f5-6bf15c9b3a88" />

Is that what you mean @zijianhuang?

**Comment by zijianhuang** [comment]
@bradwilson 

Firstly, NO bug found. The new feature with Parallel mode all is working perfectly with the console runner, as I can see the whole integration test suite with collection that launches backends (DB or Web API) run 3 times to 15 times faster. 

If I understand correctly, the new feature from this issue is not yet released to public beta. I could not find tag or branch 4.0.0-pre.151 that you use for closing this issue, but 4.0.0-pre.154. So I checked the tag out and built on my dev machine then imported them.
```xml
		<PackageReference Include="xunit.v3" Version="4.0.0-pre.154-dev" />
		<PackageReference Include="xunit.v3.runner.console" Version="4.0.0-pre.154-dev">
```
along with GlobalAssemblyAttributes.cs
```cs
using Xunit.Sdk; 
using Xunit.v3;

[assembly: Parallelization(Mode = ParallelMode.All)]
```

However, when running the same test suite with VS 2026 Test Explorer, I don't see performance boot, and visually the test cases under the same collection still run one by one, and the total time remain the same. So I presum the visual studio runner implemented weeks before 4.0.0-pre.154-dev may not yet support ParallelMode.All.

Currently my test project is still using `<PackageReference Include="xunit.runner.visualstudio" Version="4.0.0-pre.5">`, since I could not find an updated visual studio runner under "xunit\artifacts\packages" after running the build script of tag 4.0.0-pre.154.

**Comment by bradwilson** [comment]
> So I presum the visual studio runner implemented weeks before 4.0.0-pre.154-dev may not yet support ParallelMode.All.

It definitely does, and more importantly, Visual Studio will prefer to run with Microsoft Testing Platform over VSTest, so it's very likely irrelevant what version of `xunit.runner.visualstudio` you have, unless you have also gone out of your way to disable MTP (for example, using one of the `.mtp-off` packages).

That said...

I took my example (which is definitely using VSTest), made both tests run for 5 seconds, and then watched the behavior in Test Explorer. It takes 5 seconds and both tests run in parallel:

https://github.com/user-attachments/assets/baac46b2-c641-4184-a453-15c715533bd9

Then I did the exact same thing in MTP mode, and it behaved exactly the same.

So please open a new issue with a full repro project and repro steps. Thanks!
